### PR TITLE
Add Provider for MicrosoftGraph

### PR DESCRIPTION
**Pull request by Ridder90**
## Summary

(select one:)
* New provider

The windows live api is deprecated. I have created another provider to use the new MicrosoftGraph endpoints. 

- [ ] Tested
  - [ ] Unit tests
  - [ ] Style
- [ ] Documentation (PR # )

## Goal

What do you want to achieve with this PR?

## Description

Your description...

### CODE REVIEW

File: src/Provider/MicrosoftGraph.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        $apiUrl   = 'me/contacts?$top=50';
+        $contacts = [];
+
+        do {
+            $response = $this->apiRequest($apiUrl);
+            $data     = new Data\Collection($response);
+            if (!$data->exists('value')) {
+                throw new UnexpectedApiResponseException('Provider API returned an unexpected response.');
+            }
+            foreach ($data->get('value') as $entry) {
```
Comment (ApacheEx) [inline src/Provider/MicrosoftGraph.php]: let's use `$data->filter('value')->toArray()`

File: src/Provider/MicrosoftGraph.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                $userContact              = new User\Contact();
+                $userContact->identifier  = $entry->id;
```
Comment (ApacheEx) [inline src/Provider/MicrosoftGraph.php]: let's use `$entry->get('id') / $entry->get('displayName') / etc` - pls change all usages below

File: src/Provider/MicrosoftGraph.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                $userContact->displayName = $entry->displayName;
+                if (!empty($entry->emailAddresses)) {
+                    $userContact->email = $entry->emailAddresses[0]->address;
+                }
+                // only add to collection if we have usefull data
+                if (!empty($userContact->displayName) || !empty($userContact->email)) {
+                    $contacts[] = $userContact;
+                }
+
```
Comment (ApacheEx) [inline src/Provider/MicrosoftGraph.php]: no need blank line

File: src/Provider/MicrosoftGraph.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            if ($data->exists('@odata.nextLink')) {
+                $apiUrl = $data->get('@odata.nextLink');
+
+                $pagedList = true;
+            } else {
+                $pagedList = false;
+            }
+        } while ($pagedList);
+        return $contacts;
```
Comment (ApacheEx) [inline src/Provider/MicrosoftGraph.php]: let's add blank line before `return`

File: src/Provider/MicrosoftGraph.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        $userProfile = new User\Profile();
+
+        $userProfile->identifier    = $data->get('id');
+        $userProfile->displayName   = $data->get('displayName');
+        $userProfile->firstName     = $data->get('givenName');
+        $userProfile->lastName      = $data->get('surname');
+        $userProfile->email         = $data->get('mail');
+        $userProfile->emailVerified = $data->get('mail');
```
Comment (ApacheEx) [inline src/Provider/MicrosoftGraph.php]: it should be boolean - I think you need to make it as `$data->exists('mail')`

File: src/Provider/MicrosoftGraph.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+namespace Hybridauth\Provider;
+
+use Hybridauth\Adapter\OAuth2;
+use Hybridauth\Data;
+use Hybridauth\Exception\UnexpectedApiResponseException;
+use Hybridauth\User;
+
+/**
+ * Windows Live OAuth2 provider adapter.
```
Comment (ApacheEx) [inline src/Provider/MicrosoftGraph.php]: let's use `Microsoft Graph provider adapter.`

**Comment by ApacheEx** [comment]
Looks good, thanks a lot.

**Comment by Ridder90** [comment]
Travis mentions a connection timeout. How can we do a retry?

**Comment by ApacheEx** [comment]
Let's ignore Travis for now. Do not know when they will resolve the issue with timeout.

**Comment by bosiocarlosm** [comment]
Hi!,  Thanks for the new provider. When I implement this provider I can get the firstname and lastname but email is allways in null. I implemented a social login with other providers and it works.

In my API I seted "email" and "User.Read" permitions. Is email atribute working in the last version?

Thanks

**Pull request merged**